File: Algorithm/python/Main_1725.py
import sys

input = lambda: sys.stdin.readline().rstrip()


def main():
    N = int(input())
    arr = [0]
    for i in range(N):
        arr.append(int(input()))
    arr.append(0)
    stack = [0]
    result = 0

    for i in range(1, N + 2):
        while arr[stack[-1]] > arr[i]:
            height = arr[stack.pop()]
            width = i - stack[-1] - 1
            result = max(result, height * width)
        stack.append(i)

    print(result)


if __name__ == '__main__':
    main()

File: Algorithm/src/Main_1725.java
import java.io.*;
import java.util.ArrayList;
import java.util.List;
import java.util.Stack;
import java.util.StringTokenizer;

public class Main_1725 {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    static StringTokenizer st;

    public static void main(String[] args) throws IOException {
        int N = Integer.parseInt(br.readLine());
        List<Integer> list = new ArrayList<>();
        list.add(0);

        for (int i = 0; i < N; i++) list.add(Integer.parseInt(br.readLine()));
        Stack<Integer> stack = new Stack<>();
        list.add(0);
        stack.add(0);

        int result = 0;

        for (int i = 1; i < N + 2; i++) {
            while (list.get(stack.peek()) > list.get(i)) {
                int height = list.get(stack.pop());
                int width = i - stack.peek() - 1;
                result = Math.max(result, height * width);
            }
            stack.add(i);
        }
        bw.write(result + "");
        bw.close();
    }
}
